src/v/security/gssapi_authenticator.cc

```diff
@@ -196,9 +201,12 @@ ss::future<result<bytes>> gssapi_authenticator::authenticate(bytes auth_bytes) {
       });
 
     _state = res.state;
-    if (_state == state::complete) {
-        _principal = co_await _worker.submit(
-          [this]() { return _impl->principal(); });
+    if (_state == state::complete || _state == state::failed) {
+        _principal = co_await _worker.submit([this]() {
+            auto principal = _impl->principal();
+            _impl->reset();
+            return principal;
+        });
```

Comment:
I don't know if I like getting the principal from `_impl` if we're in the failed state, but I don't have any real reason to say this is a problem.

Reply:
Yeah, it should never have been assigned, so it's just default constructed.  It's probably clearer like this?
```c++
    if (_state == state::complete) {
        _principal = co_await _worker.submit([this]() {
            auto principal = _impl->principal();
            _impl->reset();
            return principal;
        });
    } else if (_state == state::failed) {
        co_await _worker.submit([this]() { _impl->reset(); });
    }
```

Reply:
Done.